cli/src/main/java/org/jboss/as/cli/impl/CLIModelControllerClient.java

```diff
@@ -297,8 +301,22 @@ public void ensureConnected(long timeoutMillis) throws CommandLineException {
                 }
 
                 if (System.currentTimeMillis() - start > timeoutMillis) {
-                    throw new CommandLineException("Failed to establish connection in " + (System.currentTimeMillis() - start)
-                            + "ms", ioe);
+                    if (timeoutOccured) {
+                        throw new CommandLineException("Failed to establish connection in " + (System.currentTimeMillis() - start)
+                                + "ms", ioe);
+                    } else {
+                        timeoutOccured = true;
+                    }
+                }
+                // Only propagate RedirectException at the end of timeout.
```

Comment:
@bstansberry, if this becomes an issue, we can change this. I first implemented a logic based on the protocol. We can retrieve it from the channelConfig and compare wth the protocol received in the redirect exception.
I although thought that it was a corner case. It would have been problematic for reload.